ELPP-3621 Downstream service sample: emit one event in response to article

giorgiosironi opened the pull request:
For #29 

See README for how much is implemented in this PR and what will follow.

Sample log:
```
downstream-sample_1                | Receiving: {'eventId': '7d5de096-6338-11e8-83a9-0242ac1a000c', 'happenedAt': '2018-05-29T12:04:50+00:00', 'aggregate': {'service': 'article-store', 'name': 'article-version', 'identifier': '0065'}, 'type': 'article.version.post.started', 'data': {'article_version': 1}}
downstream-sample_1                | Publishing: {'runId': None, 'type': 'downstream.crossref.started', 'message': 'We are happy to receive this paper and have put it into a queue', 'eventId': '72d65f3a-6339-11e8-b721-0242ac1a000a', 'happenedAt': '2018-05-29T12:11:42+00:00'}
downstream-sample_1                | Receiving: {'eventId': '7d6ef94e-6338-11e8-83a9-0242ac1a000c', 'happenedAt': '2018-05-29T12:04:50+00:00', 'aggregate': {'service': 'article-store', 'name': 'article-version', 'identifier': '0065'}, 'type': 'article.version.post.completed', 'data': {'article_version': 1}}
downstream-sample_1                | Publishing: {'runId': None, 'type': 'downstream.crossref.started', 'message': 'We are happy to receive this paper and have put it into a queue', 'eventId': '72d8925a-6339-11e8-b721-0242ac1a000a', 'happenedAt': '2018-05-29T12:11:42+00:00'}
```

Commit:
Draft of manage.py, output buffering is in place

Commit:
test.py has been removed

Commit:
Remove output buffering debug code

Commit:
Move channel orchestration to Worker.start

Commit:
Simplify leftover command

Commit:
Simplify leftover command

Commit:
Publish one event in response to article receival

Commit:
Configure using RABBITMQ_* environment variables

Commit:
Rename function to ensure_queue

Commit:
Make ensure_queue more generic

Commit:
Specify aggregate of events

Commit:
Add UUID1 generation

Commit:
Reflect on new exchange

Commit:
Add happenedAt generation

Commit:
Document key specification in README

seanwiseman commented:
Seem `__pycache__` directory is being commited

Commit:
Remove __pycache__

thewilkybarkid commented on downstream-sample/downstream_sample/settings.py:
```diff
@@ -0,0 +1,50 @@
+        # create queue, will skip if exists
+        channel.queue_declare(queue=queue_name, durable=True)
+        # bind queue to exchange, will skip if already bound
+        channel.queue_bind(exchange=ARTICLE_EXCHANGE_NAME, queue=queue_name)
```
Falls over at the moment as it doesn't exist.

seanwiseman replied:
Trying to bind before the exchange has been created.

I came up against this kind of race condition when doing the custom workflow experiment. In that project there was x5 queues each with a different exchange. My solution was to declare the queues and exchanges in a conf file (probably settings.py in this case?):

```
DEFAULT_QUEUES = {
    ACTIVITY_RESULT_QUEUE: [
        ACTIVITY_RESULT_EXCHANGE
    ],
    DECISION_RESULT_QUEUE: [
        DECISION_RESULT_EXCHANGE
    ],
    SCHEDULED_ACTIVITY_QUEUE: [
        SCHEDULED_ACTIVITY_EXCHANGE
    ],
    SCHEDULED_DECISION_QUEUE: [
        SCHEDULED_DECISION_EXCHANGE
    ],
    WORKFLOW_STARTER_QUEUE: [
        WORKFLOW_STARTER_EXCHANGE
    ]
}
```
Then decorate any entrypoint function with:

```
def setup_exchanges_and_queues(func) -> Callable[..., None]:
    """Setup required queues and exchanges on target broker.
    If they exist already they will be skipped.
    :return:
    """
    @wraps(func)
    def wrapper():
        with get_channel() as channel:
            for queue_name in DEFAULT_QUEUES:
                channel.queue_declare(queue=queue_name, durable=True)

                for exchange in DEFAULT_QUEUES[queue_name]:
                    channel.exchange_declare(exchange=exchange, exchange_type='fanout', durable=True)
                    channel.queue_bind(exchange=exchange, queue=queue_name)

        return func()

    return wrapper
```
Example usage:
```
@setup_exchanges_and_queues
def main():
    while True:
        try:
            with get_channel() as channel:
                print('Worker running...')
                print('[*] Waiting for Messages. To exit press CTRL+C')

                channel.basic_consume(message_handler(activity_handler), queue=SCHEDULED_ACTIVITY_QUEUE, no_ack=False)

                channel.start_consuming()

        except (ChannelClosed, ConnectionClosed, IncompatibleProtocolError) as err:
            # lost connection
            logger.exception('Lost connection... waiting before retry')
            time.sleep(5)
        except KeyboardInterrupt:
            channel.stop_consuming()
            break
```

This stopped me having to worry about ensuring queues and exchanges within the actual workers/listeners etc.

giorgiosironi replied:
Yes, had to test with a manually created one.

As much as exchanges can be moved around, it's safe to say an `articles` exchange, or any other the Article Store would publish to, won't be created by this project. So in this case I'd go for waiting for the `articles` exchange to appear in `get_queue`.

In the context of this project being extracted into its own repository, a dummy exchange to subscribe to should be added for development and testing purposes.

giorgiosironi replied:
Now waits using the `wait_exchange` decorator.

thewilkybarkid commented on downstream-sample/downstream_sample/settings.py:
```diff
@@ -0,0 +1,50 @@
+DOWNSTREAM_QUEUE_NAME = 'downstream-sample'
+
+CREDENTIALS = pika.PlainCredentials(USER, PASSWORD)
+PARAMS = pika.ConnectionParameters(host=HOST, credentials=CREDENTIALS)
```
Could we use `URLParameters` to simplify the environment variables (down to one).

giorgiosironi replied:
I was wondering if there was an equivalent to the PHP driver's single URL

seanwiseman replied:
For reference, this was done in [#32](https://github.com/libero/walking-skeleton/pull/32) [here](https://github.com/libero/walking-skeleton/blob/5d3e0d65dcbc75e00d9f61cba619902a06321d6f/article-store/events/conf.py#L8).

thewilkybarkid commented on downstream-sample/README.md:
```diff
@@ -0,0 +1,27 @@
+
+Implemented:
+
+- emit an `article-delivery` event upon successful receival of the `article` event
```
We're going to need to match these up between the Article Store, Dashboard and this.

The Dashboard currently knows `article.published`: https://github.com/libero/walking-skeleton/blob/9bb31a58dd2c783521d8f15dffa47b9ccecdff82/dashboard/translations/messages.en.yaml#L9-L10

(It'll display the raw name otherwise.)

Maybe this should listen for `article.published`, emit `article.crossref.started` and later `article.crossref.finished`?

thewilkybarkid replied:
Just seen https://github.com/libero/walking-skeleton/blob/89b9fee5df3e2e85c76aa5b34057e89131f4df74/downstream-sample/downstream_sample/settings.py#L17

This is https://github.com/libero/walking-skeleton/pull/28#discussion_r189177476.

giorgiosironi replied:
Let's move there, general discussion

giorgiosironi commented:
I'll be going on implementing feedback on Wednesday, assuming the general framework of #28 is confirmed.

Commit:
Merge remote-tracking branch 'origin/master' into downstream_sample

Commit:
Substitute TODO with more pressing concern

Commit:
Simplify event schema

Commit:
Use RABBITMQ_URL for configuration

Commit:
Remove comment referring to the wrong exchange

Commit:
Implement wait for exchange presence

Commit:
Merge remote-tracking branch 'origin/master' into downstream_sample

Commit:
Merge remote-tracking branch 'origin/master' into downstream_sample

Commit:
Make runId optional until implemented upstream

Commit:
Change type to use dots for hierarchy

Commit:
Updated README with implemented scenario

giorgiosironi commented:
Integrated with article store from `master` (only issue the missing `runId`), will also check the dashboard after it has changed to the new format.

Commit:
Merge remote-tracking branch 'origin/master' into downstream_sample

giorgiosironi commented:
Merged `master`, but the article store doesn't start (no logs are written) as requests to the container's `8000` port are refused despite the `python manage.py runserver` process being running.

giorgiosironi commented:
It starts when running with `docker-compose up article-store`, but `docker-compose up -V` reproduces the problem for me.

giorgiosironi commented:
Non-deterministic problem.

giorgiosironi commented:
Some manual tinkering lead to:
```
Run 4df2280c-67f8-11e8-9208-0242ac1a0010
    event.article.version.post.started
    Jun 4, 2018, 1:07:59 PM
    event.downstream.crossref.started
    We are happy to receive this paper and have put it into a queue
    Jun 4, 2018, 1:07:59 PM
    event.article.version.post.completed
    Jun 4, 2018, 1:07:59 PM
    event.downstream.crossref.started
    We are happy to receive this paper and have put it into a queue
    Jun 4, 2018, 1:07:59 PM
```

Commit:
Avoid buffering in Python server logs

Commit:
Wait for postgres in article-store startup

Partial fix for #44: the Django server should crash or wait and retry,
not hang.

Commit:
Ensure creation of downstream-sample exchange on bootstrap

giorgiosironi commented:
Had to [manually bind the queue](http://localhost:8083/#/queues/%2F/dashboard) to the second exchange however: https://github.com/php-amqplib/RabbitMqBundle/issues/367

Commit:
Swtich dashboard queue binding to multiple-exchange format

giorgiosironi commented:
Thought about adding `depends_on: - downstream-sample` for the `dashboard-events`, but it doesn't have that for the `articles` exchange and it's just a startup order (or build time) dependency, so wouldn't add any value.

thewilkybarkid committed:
Merge branch 'master' into downstream_sample

thewilkybarkid commented:
Can see it working, but seems to respond to any article event? I see two `downstream.crossref.started` events: one after `article.version.post.started` and one after `article.version.post.completed`. Would only expect to see the latter.

seanwiseman commented:
Successfully receiving the downstream events locally.

Are we going to have some sort of filtering on the received events here? e.g. only act on type `article.version.*.completed` or something along those lines? I know its not strictly important for demo purposes but with its current implementation, once you add in the work flow steps it will be firing off `downstream.crossref.started ` about 10+ times when realistically you would only want to trigger downstream once the initial article 'processing' has completed.

giorgiosironi commented:
I'll put a filter on `article.version.post.completed` that seems the correct integration point.

thewilkybarkid committed:
Filter messages

thewilkybarkid committed:
Merge branch 'master' into downstream_sample

thewilkybarkid reviewed (approved):
(If you're happy with my changes.)

giorgiosironi commented:
Took a bit to understand the flow of routing keys but :+1: feel free to merge when conveniente.
